pkg/storage/leveldbstore/transaction.go

```diff
@@ -7,12 +7,130 @@ package leveldbstore
 import (
 	"errors"
 	"fmt"
+	"sync"
 
 	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storage/storageutil"
+	"github.com/google/uuid"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/opt"
 )
 
+// PendingTxNamespace is the namespace used for in-progress reverse
+// operation transactions that has not been committed or reverted yet.
+const PendingTxNamespace = "pending-tx"
+
+var _ storage.Item = (*pendingTx)(nil)
+
+// pendingTx is a storage.Item that holds a batch of operations.
+type pendingTx struct {
+	storage.Item
+
+	val *leveldb.Batch
+}
+
+// Namespace implements storage.Item.
+func (p *pendingTx) Namespace() string {
+	return PendingTxNamespace
+}
+
+// Unmarshal implements storage.Item.
+func (p *pendingTx) Unmarshal(bytes []byte) error {
+	p.val = new(leveldb.Batch)
+	return p.val.Load(bytes)
+}
+
+// Recovery attempts to recover from a previous crash
+// by reverting all uncommitted transactions.
+func Recovery(store storage.Store) error {
+	batch := new(leveldb.Batch)
+
+	err := store.Iterate(storage.Query{
+		Factory:      func() storage.Item { return new(pendingTx) },
+		ItemProperty: storage.QueryItem,
+	}, func(r storage.Result) (bool, error) {
+		if err := batch.Replay(r.Entry.(*pendingTx).val); err != nil {
+			return true, fmt.Errorf("unable to replay batch for %s: %w", r.ID, err)
+		}
+		batch.Delete([]byte(r.ID))
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("leveldbstore: iteration failed: %w", err)
+	}
+
+	if err := store.(*Store).db.Write(batch, &opt.WriteOptions{Sync: true}); err != nil {
+		return fmt.Errorf("leveldbstore: unable to write batch: %w", err)
+	}
+	return nil
+}
+
+var _ storage.TxRevertOpStore[[]byte, []byte] = (*diskTxRevertOpStore)(nil)
+
+// diskTxRevertOpStore is a storage.TxRevertOpStore
+// that stores revert operations on disk.
+type diskTxRevertOpStore struct {
+	id       []byte // Unique identifier for this transaction.
+	db       *leveldb.DB
+	mu       sync.Mutex // mu protects the batch.
+	batch    *leveldb.Batch
+	revOpsFn map[storage.TxOpCode]storage.TxRevertFn[[]byte, []byte]
+}
+
+// Append implements storage.TxRevertOpStore.
+func (s *diskTxRevertOpStore) Append(op *storage.TxRevertOp[[]byte, []byte]) error {
+	if s == nil {
+		return nil
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var errs error
+	if fn, ok := s.revOpsFn[op.Origin]; !ok {
+		errs = errors.Join(errs, fmt.Errorf(
+			"revert operation %q for object %s not found",
+			op.Origin,
+			op.ObjectID,
+		))
+	} else if err := fn(op.Key, op.Val); err != nil {
+		errs = errors.Join(errs, fmt.Errorf(
+			"revert operation %q for object %s failed: %w",
+			op.Origin,
+			op.ObjectID,
+			err,
+		))
+	}
+	if errs != nil {
+		return errs
+	}
+
+	return s.db.Put(s.id, s.batch.Dump(), &opt.WriteOptions{Sync: true})
```

Comment:
We should see if we can get away with the sync write here. If we have to do it on every append, it will be slow as shit. Or we can have some background goroutine which keeps doing from time to time. For the revert, it should be fine.

Reply:
It would be better to bench this first.